lm_eval/tasks/mmmu_pro/utils.py

```diff
@@ -0,0 +1,228 @@
+import ast
+import random
+import re
+
+import numpy as np
+
+
+random.seed(42)
+
+# ============ Direct lm_eval utility functions. Copied from mmmu/utils.py ============ 
+# Link: https://github.com/EleutherAI/lm-evaluation-harness/blob/main/lm_eval/tasks/mmmu/utils.py
+
+
+# From https://github.com/MMMU-Benchmark/MMMU/blob/main/mmmu-pro/prompts.yaml
+MULTI_CHOICE_EXAMPLE_FORMAT = """{}
+
+{}
+
+Answer with the option letter from the given choices directly. The last line of your response should be of the following format: 'Answer: $LETTER' (without quotes) where $LETTER is one of the options."""
+
+
+START_CHR = "A"
+
+
+def doc_to_image(doc):
+    # get formatted prompt (incl. multi-choice options) pre-<image {i}> reformatting
+    input_text = _doc_to_text(doc)
+    # locate <image {i}> instances in input
+    image_placeholders = [
+        img.replace(" ", "_").replace("<", "").replace(">", "")
+        for img in re.findall("<image [1-7]>", input_text)
+    ]
+
+    # collect visuals (can have dupes of a given image or be out of order)
+    # E.g. validation_Math_19 contains <image 1> and <image 2> but seen as [<image 1>, <image 2>, <image 1>, <image 1>, <image 2>]
+    visuals = [doc[img] for img in image_placeholders]
+
+    return visuals
+
+
+def doc_to_text(doc):
+    """Get the prompt for a given document."""
+
+    prompt = _doc_to_text(doc)
+
+    for i in range(1, 8):
+        # replace <image {i}> with <image>. TODO: check this is always the right decision incl. for non-HF models
+        prompt = prompt.replace(f"<image {i}>", "<image>")
+
+    return prompt
+
+
+def _doc_to_text(doc):
+    """Helper--get the prompt for a given document but DO NOT yet replace <image {i}> with <image>."""
+    choices_str = ""
+
+    for i, choice in enumerate(ast.literal_eval(doc["options"])):
+        # add (A) {choice1}\n , (B) {choice2}\n , and so on
+        # to create the list of formatted choices in the prompt
+        choices_str += f"\n({chr(ord(START_CHR) + i)}) {choice}"
+
+    choices_str = (
+        choices_str.lstrip()
+    )  # remove the extraneous prepended \n that we added
+
+    prompt = MULTI_CHOICE_EXAMPLE_FORMAT.format(doc["question"], choices_str)
+
+    return prompt
+
+
+def process_results(doc, results):
+    # multichoice logic
+    option_strs = ast.literal_eval(doc["options"])
+    index2ans, all_choices = get_multi_choice_info(option_strs)
+
+    pred = parse_multi_choice_response(results[0], all_choices, index2ans)
+    # print(pred, all_choices, index2ans)
+    is_correct = eval_multi_choice(doc["answer"], pred)
+
+    return {
+        "acc": float(is_correct),
+    }
+
+# ============ MMMU-Pro Vision Variant lm_eval utility functions, written by hand (extrapolated from the above) ============ 
```

Comment:
Whew did you have to reverse engineer this? How long did this take to write and verify? Nice work powering through getting this hooked up.

In an ideal world we would check that the final model response agrees with the official repos for some prompts, but in the interest of time if the aggregates match up for multiple models we can unblock MLEs. There are so many details that could go wrong with this kind of intricate code 😨 .